### ISSUE TITLE
[Feature Request] Write restart checkpoint on error

### ISSUE THREAD

**Issue by Golui** [lammps/lammps#3955]
**Summary**

Troubleshooting simulations often requires at the very least the final frame of a simulation. This feature would facilitate saving the coordinates or whole simulation checkpoints whenever LAMMPS encounters an error (instability, missing atoms, invalid command in the input etc.).

**Detailed Description**

When troubleshooting simulations, it is often handy to have, at the very least, the final frame of a trajectory on hand. Often this involves creating a `dump` that writes to a trajectory every frame. This is however not an option for diagnosing simulations that, for whatever reason, fail after long simulation times. While most of this can be fixed by just creating more robust protocols and inputs, it is the times when one does not take the necessary precautions that such data becomes invaluable.

As such, I propose creating a feature to write a restart file whenever lamps encounters an error. A sort of "panic" checkpoint.  I propose saving the entire state, since saving coordinates directly is not so straightforward in the general case. Moreover, saving a checkpoint would allow users to process it themselves, via creating a new input file that loads it, processes it, and, for example, dumps coordinates of the current frame.

This also has the added benefit that if a user has a typo in their input file inside of a command after the MD run (say, in the `write_restart` command after a run), and since the `write_restart` command does _not_ save the restart file if the last timestep is not a multiple of `N`, some frames might get lost.

I see this feature being on by default, and disabled only when explicitly turned off. Alternatively, perhaps extending hte behaviour of the `restart` command to allow it to both save final restart files for runs, and on errors during the simulation (both controlled via extra (keyword?) arguments to the command) would be in order.

**Further Information, Files, and Links**

I am still fairly new to LAMMPS, so please forgive me if this is already achievable. I did not see any Issues, Pull Requests, mention in the documentation, or information from much more experienced colleagues about such functionality. 

I am willing to implement it myself, but would require some guidance to ensure a quality feature comes out of this.

**Comment by akohlmey** [comment]
@Golui this all sounds great only that there is a significant problem: by the time you detect an error condition, it is too late. The system *is* already in a bad state, e.g. you are already missing atoms or have NaN for coordinates or other, similar problems. Writing a restart is meaningless at this point, since you cannot continue from there. In addition, usually, you cannot expect that MPI communication is usable at that point but that is required for writing a restart. There have been discussions in the past about using exception handling, but also that has a problem, since exceptions are not propagated across MPI ranks, so it can only be done in a reasonable way for serial execution. The LAMMPS GUI implementation is set up this way, so that errors in the LAMMPS instance can be detected and will not crash the application itself, instead the LAMMPS instance is discarded and a new one created. However, running in serial is not the common case for LAMMPS and implementing a serial-only feature into a program meant to be used for *many* parallel processes is not that helpful.

The only meaningful prevention for a total loss of data is to use the existing restart command with a suitable restart interval and possibly write_restart at the end of each run. restart is best used with two files to alternate between them or with a file name that includes "*" so a new restart is written every time to write out restarts are regular intervals while the system is still valid.

Your discussion of the the "write_restart" command functionality is not correct either. It *will* instantly write one restart file. You seem to be mixing it up with the "restart" command.

In addition, the use case scenario that you are discussing (e.g. typos in the input), can be easier handled by running a smaller test system or using the [-skiprun command line flag](https://docs.lammps.org/Run_options.html#skiprun) or both. The fact that crashes can happen at the wrong time when not following best practices should be a (painful) reminder to *always* follow best practices.

In general, it looks to me that with reasonably careful testing and planning of simulations the problematic scenario is a non-issue. The use case would be during testing and not during production simulations, since the scenarios you describe should not happen for properly set up simulations. Beyond that, the straightforward way to do post-mortem analysis would be to enable core dumps and use a debugger to inspect them, but that can be *very* impractical for parallel runs.

In summary, what you propose either cannot be implemented for most cases and would have practically no value.

**Comment by Golui** [comment]
@akohlmey Thanks for the quick reply and the tip about `-skiprun`! Did not consider to use it; will add it to my workflow.

You are correct - in my initial comment, the statement "[...] and since the `write_restart` command [...]" was indeed supposed to refer to the `restart` command. Sorry for the mix-up.

This is coming from a long-time GROMACS user, so please take this with a grain of salt. In GROMACS, when _some_ errors occur (such as bad physics) [a few post-mortem PDB files are created](https://gitlab.com/gromacs/gromacs/-/blob/f47a5a1b0099f295539169712defd46241fd381f/src/gmxlib/update.c#L625), for the frame before and after the incident. I find these to be useful in diagnosing. While LAMMPS gives you the indices of atoms where a problem occurred, they are not always super informative. For example, I have seen a new member of our group struggle with errors relating to the "SHAKE" fix, where the actual problem was atoms escaping out of a non-periodic dimension in the box. Visual inspection combined with the index information would have made the troubleshooting much easier, so I don't really understand why you consider this having "practically no value".

**Comment by akohlmey** [comment]
> for the frame before and after the incident.

I suspect this is only possible because gromacs using leapfrog instead of velocity Verlet where you need to keep a copy of the previous coordinates. As I already stated, by the time there is an error condition detected in LAMMPS, it is too late and you cannot really depend on MPI communication functioning.

> Visual inspection combined with the index information would have made the troubleshooting much easier, so I don't really understand why you consider this having "practically no value". 

this is quite a corner case and you cannot output information that you don't have (anymore).

It has been about 20 years since I have last used gromacs seriously, but I recall to have those "emergency PDB" files output only when for example the potential energy became very high. This scenario can also already be handled in LAMMPS using the "fix halt" command (and a write_restart or write_data or write_dump after the run).
In general, when comparing gromacs and LAMMPS you have to understand that while gromacs is more a monolithic application, LAMMPS is more of a toolkit with lots of bits and pieces that can be added (or not). This flexibility comes at a price, i.e. each feature has to take care of its own contribution to initializing, communicating, restarting and so on. 

There is an ongoing project (with rather slow progress due to lack of manpower) to improve error messages. See, e.g. https://www.lammps.org/workshops/Aug22/project1/ Previously the philosophy was that detailed error message create too much maintenance work (as documentation would need to be synchronized and errors explain), so LAMMPS would print just the equivalent of "there was an error" or "your input was invalid" with a line number of the source code where the error happened. And it was expected that LAMMPS users would be sufficiently proficient in C/C++ to deduce the actual cause of the problem from the source code. After acknowledging that this expectation is no longer valid for most LAMMPS users (especially as the user base is growing and the average proficiency in C/C++ programming dropping at the same time), changes that improve error reports along the given examples from the 2022 workshop will be welcome.

However, I do not see how what you propose can be implemented in a meaningful way and without negatively impacting parallel scaling behavior (something that gromacs traditionally is struggling with more than LAMMPS, but makes up for it with faster force kernels (on the CPU)). 

Feel free to prove me wrong, however, I don't know how to provide guidance for something I don't believe can be done.

**Comment by Golui** [comment]
I absolutely understand that there are enormous architectural differences between GROMACS and LAMMPS, and I do not expect LAMMPS to have all the same functionality of GROMACS and vice-versa. At this point it was more of a question out of curiosity, knowing that another package can do it. Obviously, a massive architectural overhaul for such a trivial feature is out of the question.

At this point I will give up on this, unless I can come up with a very clever way of doing that. 

Thank you for the quick and very detailed replies.
